### ISSUE TITLE
No way to control which dns zone is used

### ISSUE THREAD

**Issue by mortenn** [caddy-dns/azure#9]
I have a split DNS deployment where subdomain.example.com is a private dns zone and example.com is a public dns zone.
When I add a route for service.subdomain.example.com Caddy fails to add a record to subdomain.example.com, which is the wrong dns zone to attempt to update.

I was unable to find a setting for which zone to use, so this seems to not work at present.

**Comment by mortenn** [comment]
I also tried setting AZURE_DNS_ZONE_FQDN after reading the libdns readme, but it had no effect

**Comment by kurokobo** [comment]
Thank you for raising the issue. At this time, private DNS zone is not supported.

I may not fully understand the goals you want to achieve, but is your scenario that the CA performing the DNS-01 challenge is a private CA located on a VNet, rather than a public CA like Let's Encrypt?

**Comment by mortenn** [comment]
yeah, I don't want it to use the private dns zone, but it does because the private dns zone is x.y.z when the fqdn is foo.x.y.z
I managed to work around it by overriding the challenge url to foochallenge.y.z and add a cname for the real challenge url

**Comment by jasonfare** [comment]
Not sure what changed as my configuration used to work just fine for almost a year against the primarydomain.tld.  Recent changes now require to have a dedicated subdomain with NS Delegations in place.  

If zone is missing, it does not attempt to create one.

This is how I had to get around this:
1. Create a new Public DNS zone called _acme-challenge.subdomain.primarydomain.tld.   
   a.  If doing through the portal, you only put the _acme-challenge.subdomain as the name being added and select the "This zone is a child of an existing zone already hosted in Azure DNS."  Select appropriate subscription and parent zone.
   b. If doing through CLI/PS, create new FQDN zone and in primarydomain.tld, create NS entries back to Azure DNS for FQDN (_acme-challenge.subdomain.primarydomain.tld)
4. Delegate DNS Zone Contributor to the service principal/managed identity over the newly created acme challenge zone 
5. Restart Caddy (it then was able to update the correct location.)

**Comment by kurokobo** [comment]
I'd like to clarify that, as a premise, Caddy is the one that determines which zone needs to be updated, and this module simply uses the zone information provided by Caddy as is.
If we want to change the zone in which records are updated, we might need to consider adding a configuration option that forcefully overrides the zone name.

@mortenn 
Sorry for asking this after some time has passed.

Based on your requirements, which of the following are you expecting Caddy to do?

- Add the `_acme-challenge` record to your private `subdomain.example.com` zone
- Add the `_acme-challenge.subdomain` record to your public `example.com` zone

@jasonfare 
> Recent changes now require to have a dedicated subdomain with NS Delegations in place.

Is this issue related to the OP's Private DNS zone?
Also, is this because the version of `dns.providers.azure` has been updated?

If possible, could you share a sample of the configuration (a caddyfile snippet, with the domain name replaced) that used to work but has stopped working recently?

**Comment by mortenn** [comment]
The latter, updating the private dns zone does not work, as that is private and the acme server won't resolve the challenge.

For my usecase, the workaround outlined above using a manual challenge url is sufficient.

**Comment by kurokobo** [comment]
Since there is a workaround and there don't seem to be any strong requirements to natively support private DNS zones, I'll close this issue. If there is strong demand for this requirement, please feel free to open a new issue.

**Comment by mortenn** [comment]
The point was not to support private dns zones, in fact, the opposite; it tried using a private zone by default and thus failing.
But I think the workaround should work fine for most anyone facing the same issue.
